reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala

```diff
@@ -56,9 +57,15 @@ final class StreamUnicastPublisher[F[_]: Async, A](
 }
 
 object StreamUnicastPublisher {
+  @deprecated("Use overload which takes only the stream and returns a Resource", "3.4.0")
   def apply[F[_]: Async, A](
       s: Stream[F, A],
       dispatcher: Dispatcher[F]
   ): StreamUnicastPublisher[F, A] =
-    new StreamUnicastPublisher(s, dispatcher)
+    new StreamUnicastPublisher(s, dispatcher, dispatcher)
+
+  def apply[F[_]: Async, A](
+      s: Stream[F, A]
+  ): Resource[F, StreamUnicastPublisher[F, A]] =
+    (Dispatcher.sequential[F], Dispatcher.sequential[F]).mapN(new StreamUnicastPublisher(s, _, _))
```

Comment:
Since we can now have local `Dispatcher`s with special sequential semantics, it seems like we should deprecate the pattern of asking for (presumably more global) dispatchers in favor of building our own dispatchers.

Furthermore, since we now rely on the sequential `Dispatcher` to achieve the correct semantics, exposing it as an argument is not good design.

Reply:
I'm sorry to come in to this so late in the game after this has merged, but we use this in a Play application to stream CSV files from an fs2 `Stream`; the Play API requires a class that's hard to construct except by using a `StreamUnicastPublisher` but it doesn't give us the opportunity to call close. In previous discussions on discord, I was advised to create a `Dispatcher` for the application and pass it in; we can guarantee the provided `Dispatcher` gets closed at the end.

I see (superficially) the point of your comment here and the rationale behind deprecating the API that takes a provided `Dispatcher`, but I don't see how we can migrate to the non-deprecated API without leaking `Dispatcher` instances. I will try to find you on discord sometime to talk about it a little more.

Reply:
@mtomko no problem, thanks for the comment.

> but I don't see how we can migrate to the non-deprecated API without leaking `Dispatcher` instances.

I need to think about it more, but so long as you are careful to cancel the subscription, I don't think there will be any leak here. The `Resource` surrounding a `Dispatcher` is intended to make sure that any tasks it started are canceled when the resource closes. If a `Dispatcher` has no running tasks and it goes out of scope, I am fairly certain that the GC will completely clean it up.

Also @BalmungSan is currently working on some usability improvements that might help your usecase )

Reply:
I'm trying to trace through the code, but it means digging deeply into akka streams and Play. The code _we've_ written just creates the source and passes it to Play's `Ok` to construct a result:

```scala
  def streamToClientAsCsvFile(
      enumerator: Stream[IO, Row],
      dispatcher: Dispatcher[IO],
      filename: String
  ): Result = {
    val publisher = StreamUnicastPublisher(enumerator.through(rowToCsv), dispatcher)
    val source = Source.fromPublisher(publisher)

    Ok.chunked(source)
      .withHeaders(CONTENT_DISPOSITION -> s"attachment; filename = $filename")
      .as("text/csv")
  }
```

I would like to hope that Play is smart enough to recognize when the connection is closed and close the `Source`.

Reply:
Ps, thanks for your response, it's really appreciated!

Reply:
@mtomko #3107 is the PR @armanbilge mentioned. However, I am not exactly sure how it helps in your use case.

Reply:
> However, I am not exactly sure how it helps in your use case.

@BalmungSan @mtomko the new `stream.subscribe(...)` syntax that Luis is adding in https://github.com/typelevel/fs2/pull/3107 will enable you to bring back the global `Dispatcher`.